src/main/java/uk/gov/hmcts/reform/wataskmanagementapi/services/TaskManagementService.java

```diff
@@ -329,6 +331,21 @@ public void cancelTask(String taskId,
         requireNonNull(accessControlResponse.getUserInfo().getUid(), USER_ID_CANNOT_BE_NULL);
         PermissionRequirements permissionsRequired = PermissionRequirementBuilder.builder().buildSingleType(CANCEL);
 
+        String userId = accessControlResponse.getUserInfo().getUid();
+        String email = accessControlResponse.getUserInfo().getEmail();
+
+        if (isGranularPermissionFeatureEnabled(userId, email)) {
+
+            final Optional<TaskResource> taskResource = getTaskById(taskId);
```

Comment:
You can optimise this by calling verifyRoleAssignments with CANCEL or CANCEL_OWN permission and then using the returning task resource, you can check the assignee if the only permission that task has got is CANCEL_OWN. By doing that you can cut off extra db call done get getTaskById method

Reply:
Logic rewritten to avoid extra db call.